### PR TITLE
Fix Settings UI silently dropping failed writes (no user-facing error)

### LINKED ISSUE #54617: Settings UI silently drops failed writes (no user-facing error)
### Reproduction steps

1. Open your user settings file path.
2. Make `settings.json` read-only (Windows: set file attribute read-only, or deny write permission).
3. Open Zed **Settings** UI and change a value (for example Theme Name, Font Family, or a toggle).
4. Observe UI behavior and then reopen settings / file contents.

**Observed:** No visible error toast/banner; failure is only logged internally. The setting may appear unchanged or revert without clear explanation.

### Summary

In the Settings UI, most field updates call `update_settings_file(...)` and then immediately do `.log_err()` with TODO comments. If writing settings fails (permissions, file lock, bad path, I/O error), the failure is only logged and not surfaced to the user.

That creates a silent-failure workflow: user changes a setting, but Zed gives no actionable error in UI.

### Current behavior

- Settings write errors are handled via `.log_err()` in many UI controls.
- Comments already indicate this is not ideal (`todo(settings_ui) don't log err`).
- User receives no direct feedback that persistence failed.

https://github.com/user-attachments/assets/ea450223-a8bb-45df-8e36-cc4eee9858f5

### Expected behavior

- Show a clear error to the user when settings persistence fails (toast/banner with reason).
- Keep UI state consistent with persisted state and provide retry guidance.

https://github.com/user-attachments/assets/de8c3a42-2573-41e2-a0b2-7e5c71f6a67d

### Zed version and system specs

Zed: v0.234.0+dev.554f5c6b65708e054ee34a12e0b8015cb9bed71f
OS: Windows
RAM: 32GB

### Attach Zed log file

<details><summary>Zed.log</summary>

```log

```

</details>

### Relevant Zed settings

<details><summary>settings.json</summary>

```json

```

</details>

### Relevant Keymap

<details><summary>keymap.json</summary>

```json

```

</details>

### (for AI issues) Model provider details

_No response_

### If you are using WSL on Windows, what flavor of Linux are you using?

None

### PR DESCRIPTION
**Pull request by galuis116**
Self-Review Checklist:

- [x] I've reviewed my own diff for quality, security, and reliability
- [x] Unsafe blocks (if any) have justifying comments
- [x] The content is consistent with the [UI/UX checklist](https://github.com/zed-industries/zed/blob/main/CONTRIBUTING.md#uiux-checklist)
- [x] Tests cover the new/changed behavior
- [x] Performance impact has been considered and is acceptable

Closes #54617

Release Notes:

- Fixed 
Settings UI now surfaces save failures directly in the UI instead of silently logging them.
When a settings write fails (for example due to file permissions, path issues, or write conflicts), Settings now shows a warning banner with details and a dismiss action.

### CODE REVIEW

**Comment by joaovictor712** [comment]
waiting to  check

**Comment by joaovictor712** [comment]
Hello @SomeoneToIgnore,  I am waiting for you to check and merge this pull request.

**Comment by probably-neb** [comment]
I ask that you please have some patience and do not @-mention maintainers without reason. SomeoneToIgnore was helpfully looking at the PR, and assigning it to me to give a thorough review as I have experience in this part of the codebase. They were not avoiding giving a review to make you wait longer

File: crates/settings_ui/src/settings_ui.rs [hunk 1]
```diff
@@ -3906,14 +3936,51 @@ fn update_settings_file(
...
+                })
+                .detach();
+            }
             Ok(())
         }
         SettingsUiFile::Server(_) => unimplemented!(),
     }
 }
 
+pub(crate) fn update_settings_file_or_notify(
```
Comment (probably-neb) [inline crates/settings_ui/src/settings_ui.rs]: small nit: I think this top level function should still be called `update_settings_file`, with the inner function just called `update_settings_file_inner`. If wrapping the inner function is the best here

File: crates/settings_ui/src/settings_ui.rs [hunk 2]
```diff
@@ -3906,14 +3936,51 @@ fn update_settings_file(
...
+
+                    if let Some(message) = message {
+                        let _ = settings_window.update(cx, |this, cx| {
+                            this.settings_write_error = Some(message);
+                            cx.notify();
+                        });
+                    }
+                })
+                .detach();
+            }
```
Comment (probably-neb) [inline crates/settings_ui/src/settings_ui.rs]: Why is the user branch also updating the error? It's wrapped in a function that does this same update, write err, notify code?

File: crates/settings_ui/src/settings_ui.rs [hunk 3]
```diff
@@ -3932,11 +3999,16 @@ impl Global for ProjectSettingsUpdateQueue {}
...
+        let (tx, mut rx) = mpsc::unbounded::<ProjectSettingsUpdateEntry>();
         let task = cx.spawn(async move |mut cx| {
             while let Some(entry) = rx.next().await {
+                let settings_window = entry.settings_window.clone();
                 if let Err(err) = Self::process_entry(entry, &mut cx).await {
                     log::error!("Failed to update project settings: {err:?}");
+                    let _ = settings_window.update(cx, |this: &mut SettingsWindow, cx| {
+                        this.settings_write_error = Some(format!("{err:#}"));
+                        cx.notify();
+                    });
```
Comment (probably-neb) [inline crates/settings_ui/src/settings_ui.rs]: here this error code is duplicated a third time

**Review by probably-neb — CHANGES_REQUESTED** [review]
This feels like the minimally invasive change to report the error, and I think we want something more comprehensive here.

I think there are two nicer ways we could structure this code.

Option A) Have a shared helper that extracts the `SettingsWindow` from the `window`, and sets the error. Then we call this in the project update, user update, and wrapper function. This is the most similar to what you've done, but a bit tidier. 

Option B) (my preferred) I think instead of duplicating this error handling code everywhere, we could restructure the user file and project file update functions to return `Task<Result<()>>`, e.g. force them to return errors (can use `Task::ready(err)` for errors that happen before submitting the update to the update queues). For the project this could be accomplished with a channel that sends the error once the update fails, and returning a task that awaits that error to arrive  on the channel and returns it. For the user file it could be accomplished similar to what you're doing now, just returning the `cx.spawn` task that awaits the update and returning the result instead of handling it.
Then the wrapper function could just `cx.spawn(|cx| { if let Err(e) = result.await { /* update error */ }}).detach()`. This feels much cohesive to me. Fundementally the user and project codepaths operate the same. They check some preconditions and do some setup that could fail, then they submit the update function to a queue - Right now they don't await/receive any information on the result of the update, but they could.

File: crates/settings_ui/src/settings_ui.rs (hunk 3 above)
Comment (joaovictor712) [reply crates/settings_ui/src/settings_ui.rs]: I fixed and pushed new commit

File: crates/settings_ui/src/settings_ui.rs (hunk 2 above)
Comment (joaovictor712) [reply crates/settings_ui/src/settings_ui.rs]: I fixed and pushed new commit

File: crates/settings_ui/src/settings_ui.rs (hunk 1 above)
Comment (joaovictor712) [reply crates/settings_ui/src/settings_ui.rs]: I fixed and pushed new commit

**Comment by joaovictor712** [comment]
Hello @probably-neb, I am waiting for you to merge this pull  request.

**Comment by joaovictor712** [comment]
Hello @probably-neb , would you like to take a time to check and merge this pull request?

**Comment by SomeoneToIgnore** [comment]
We discourage pinging maintainers, pinging them each day is bad.
I'll leave it to @\probably-neb 's to reopen and review the PR, if they still want to, after all.

**Comment by SomeoneToIgnore** [comment]
We have discussed the situation with @galuis116 and seem to have come to an agreement that pinging is bad and would not be tolerated further, hence we can proceed with the PRs when the reviewers can.

**Comment by probably-neb** [comment]
Hey @galuis116 

I've pretty much rewritten the PR from scratch. I was looking through it once you made the follow up changes, and realized it really didn't meat the quality bar we expect code to meet to merge into Zed. I considered closing the PR due to your pings of myself and `SomeoneToIgnore`, and the fact that the code was clearly AI written and of low quality, however, the change is worthwhile so I have gone about doing it myself.
I ask that in the future if you want to contribute to Zed, please write the code yourself, or use an LLM to make the code _better_, do not just ask the LLM to write the code. 
I hope my changes serve as an example of the kind of changes we look forward to seeing from the community.

**Comment by joaovictor712** [comment]
I got it. Thank you for your advice.

**Comment by probably-neb** [comment]
Hey, @galuis116. I'm going to close this now. We have given you multiple warnings to not ping maintainers, and you continue to disregard those warnings. I understand you have your own schedule, I ask that you understand we have our own schedules as well. I was going to merge my rewrite on top of yours as a matter of respect and goodwill despite the code no longer being yours, however, I feel at this point you have lost that privilege.

**Comment by joaovictor712** [comment]
@probably-neb  not fair, If you have time to close this, why did you merge this before I ping you?
Also I deleted my comment as I noticed I would disregard your warning.

**Comment by joaovictor712** [comment]
I have waited for this pull request to merge since your last commit for 5 days.

**Comment by joaovictor712** [comment]
if you did not ping you, this pull request would keep opened.

**Comment by probably-neb** [comment]
To address your points directly:

> If you have time to close this, why did you merge this before I ping you?

The PR was not in a mergeable state. I should have said so plainly after pushing my rewrite — that's on me, and I'm sorry for the ambiguity. That does not excuse what followed.

> if you did not ping you, this pull request would keep opened.

This is exactly the problem. You have continued to spam myself and `SomeoneToIgnore` trying to get us to move faster after repeated warnings to stop, be patient, and respect boundaries. You clearly haven't learned, and so the PR will stay closed
